### PR TITLE
WIP: Add ability to modify matched router parameters in middleware

### PR DESCRIPTION
**Pull request by ahopkins**
Currently, there is no way to modify the matched parameter values on the router in the middleware.

Part of the problem is that the router is called potentially multiple times in the same request. This PR **does not** attempt to fix that problem. But, instead it tries to allow both the getting and setting of matched parameter values using the existing `request.match_info`.

### CODE REVIEW

**Comment by vltr** [comment]
Oh boy this is awesome! :sunglasses:

**Comment by sjsadowski** [comment]
I like it. Let's get it in to 19.03

File: sanic/router.py [hunk 1]
```diff
@@ -391,15 +391,20 @@ def get(self, request):
         """
         # No virtual hosts specified; default behavior
         if not self.hosts:
-            return self._get(request.path, request.method, "")
+            processed = self._get(request.path, request.method, "")
```
[the remark this answers is not in the document]
Comment (ahopkins) [reply sanic/router.py]: Let's leave this one on hold until after the next release and see what happens (if anything) with the router.

**Comment by em92** [comment]
Using your branch of sanic (latest commit 4d22b6c3680d1a7d357e031fd4276c9b5c2e8e00).
Example app.
```python
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

from sanic import Sanic

from sanic.response import text

app = Sanic()

@app.middleware
def add_underscore(request):
    if request.match_info and 'username' in request.match_info:
        request.match_info['username'] = "_" + request.match_info['username']


@app.route("/test/<username>")
def http_test(request, username):
    return text(username + "\n")

app.run( host = "0.0.0.0", port = 8080 )
```

Assuming `/test/hey` will always return `_hey`, but:
```bash
$ curl http://vm1:8080/test/hey
_hey
$ curl http://vm1:8080/test/hey
__hey
$ curl http://vm1:8080/test/hey
___hey
$ curl http://vm1:8080/test/hey
____hey
$ curl http://vm1:8080/test/hey
_____hey
$ curl http://vm1:8080/test/hey
______hey
$ curl http://vm1:8080/test/hey
_______hey
```

**Comment by em92** [comment]
Disabling routing cache bug above, but it does not seem to be appliable solution:
```diff
diff --git a/sanic/router.py b/sanic/router.py
index 6110a55..a9bf94c 100644
--- a/sanic/router.py
+++ b/sanic/router.py
@@ -416,7 +416,7 @@ class Router:
         # if methods are None then this logic will prevent an error
         return getattr(route, "methods", None) or frozenset()
 
-    @lru_cache(maxsize=ROUTER_CACHE_SIZE)
     def _get(self, url, method, host):
         """Get a request handler based on the URL of the request, or raises an
         error.  Internal method for caching.
```

**Comment by ahopkins** [comment]
Thanks for catching this. We certainly do not want to bypass the route cache. 

## TODO
- [ ] Add test that hits the endpoint multiple times without unintended consequences from caching the value of `match_info`

**Comment by vltr** [comment]
I really have a small hunch that the cache may not be the real issue here (although disabling it _seems_ to fix the problem). I'll take a closer look.

**Comment by vltr** [comment]
The problem here seems to be that the `Router.get` method is called **twice** in this request lifecycle, since `match_info` is a property on the `Request` object (that calls the router again). I kind of described this request workflow problem on #1317. IMO, the `match_info` should be a object (`dict`, `list` of `namedtuple`, `immutables`, Idk, to be discussed) inside the request instance - but that would require a lot of work on the protocol, router, request and I don't know what else (described in the referenced issue) ...

**Comment by ahopkins** [comment]
@vltr I agree. This is something that belongs on the `Request`. So, while this is a worthwhile and needed feature, the hacky hoops to jump thru are probably not worth it right now. 

We need to spend some effort as a community thinking thru the router. I am putting this on hold until we decide.

**Comment by sjsadowski** [comment]
Moving to 19.6

**Comment by ahopkins** [comment]
I think we can bump this one over to 19.9.

I have some ideas that I want to tackle regarding the router and have been playing around with some implementation ideas. I think I found a better way to accomplish this and solve some of the repetitive calls to the router in Sanic.

Regardless, I am closing this PR now because I do not think this is the right approach.

**Pull request closed without merging**